feat: implement pact message

williaminfante opened the pull request:
MVP Message Pact implementation for Consumer:
- Add message pact implementation with single message support.
- Extract common broker publish function to `broker.py`
- Add new example test project `message`

used the following as references:
- https://github.com/pact-foundation/pact-message-ruby
- https://github.com/pact-foundation/pact-python/pull/119
- https://docs.pact.io/implementation_guides/jvm/consumer/groovy/#consumer-test-for-a-message-consumer

elliottmurray reviewed (commented):
So this is a pretty impressive effort but a few things I see (and I'm still processing it all!)

elliottmurray commented:
So this is a pretty impressive effort but a few things I see (and I'm still processing it all!) 

You are using some more modern Python features. Today I deprecated 3.5 and 3.4 and added 3.9 so this will work. Also I had acciedentally disabled CI on PR's. So can you rebase from master and check that the CI test bit kicks in? 

And how hard would it be to wire up your example into CI? I do it for e2e but you may need to not use broker. Nice work on the broker class btw. Had a minor question around time outs but probably get those done first.

elliottmurray commented:
I created a gist that seems to work. I can do the github workflow bit but at least have the examples Makefile working locally. 

https://gist.github.com/elliottmurray/12ebd6b784a1674697e0e00e782182cf

elliottmurray commented:
And should there be a provider sample test? I'm not up with the messaging stuff but seems like we only test consuming?

williaminfante commented on examples/message/tests/consumer/test_message_consumer.py:
```diff
@@ -0,0 +1,96 @@
+        # handler needs 'documentType' == 'microsoft-word'
+        MessageHandler(expected_event)
+
+    time.sleep(1)
```
Thanks for the review. We added `time.sleep(1)` because we needed to wait for the generation of the json pact. Without this, the test will check for the existence of pact even before it's generated. We'll have a look at progressive delays to replace the sleep.

williaminfante commented:
> I created a gist that seems to work. I can do the github workflow bit but at least have the examples Makefile working locally.
> 
> https://gist.github.com/elliottmurray/12ebd6b784a1674697e0e00e782182cf



> I created a gist that seems to work. I can do the github workflow bit but at least have the examples Makefile working locally.
> 
> https://gist.github.com/elliottmurray/12ebd6b784a1674697e0e00e782182cf

Thanks Eliott! I tried the example makefile from gist and it was working locally. I'll check later too if it still works after more PR changes.

williaminfante commented:
> And should there be a provider sample test? I'm not up with the messaging stuff but seems like we only test consuming?

We are currently working on a separate branch (possible PR soon) for supporting a message provider and its accompanying message provider tests (we're planning to place an example providing test within the `example/message/` too in that future PR). 

For this PR, we only have the consuming test.

elliottmurray commented:
All sounds very reasonable. If you get the makefile change in I'll merge and probably do a release. The others are enhancements I can work with you on. 

Great work!

williaminfante commented:
> All sounds very reasonable. If you get the makefile change (and rebase) in I'll merge and probably do a release. The others are enhancements I can work with you on.
> 
> Great work!

Thanks! We've now rebased the file, revised makefile and added the progressive delays in test.

ptsiakos77 commented:
Great job I have though a question. I can generate pacts that include matchers in with_content builder throwing T**_ypeError: Object of type Like is not JSON serializable_**. see below example
 

**_from pact.message_consumer import MessageConsumer, Provider
from pact.message_pact import MessagePact
from pact.matchers import Like, Term, Format

consumer = MessageConsumer('TestConsumer')
provider = Provider('TestProvider')

message_pact = MessagePact(consumer=consumer, provider=provider, pact_dir='spec/pacts')

(message_pact.
    given('Provider state')
    .expects_to_receive('an account mesage')
    .with_content(
        {
            'id': 'dwcvdvv',
            'uid': 'dvsfvff',
            'name': Like('Winterfell'),
            'website_url': Term(
                matcher="^(http|https):\/\/[a-z0-9]+(-[a-z0-9]+)*\.+[a-z]{2,}$",
                generate='http://winterfell.com'
            ),
            'subdomain': Like("winterfell"),
            'domain': Term(
                matcher="^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$",
                generate="winterfell.com"
            ),
            'description': Like("Lorem ipsum dolor sit amet."),
            'style': Like("minimal"),
            'translations': {
                'en': {
                    "description": Like("Lorem ipsum dolor sit amet")
                }
            }
        }
    )
    .with_metadata({'contentType': 'application/json', 'source': 'legacy_api'})
)_**


I believe that maybe the we should make a change here:
    def with_content(self, contents):
        self._insert_message_if_complete()
        self._messages[0]['contents'] = contents (Replace with self._messages[0]['contents'] = from_term(contents) #from matcher package
        return self

williaminfante commented:
> @williaminfante Great job!
> I have though a question... I can't generate pacts that include matchers in **with_content** builder throwing **TypeError: Object of type Like is not JSON serializable.**
> 
> Possible quick fix is to change L139 of **message_pact.py** with: **_self._messages[0]['contents'] = from_term(contents)_**
> importing **from_term** from **matchers.py**
> 
> See example below:
> https://gist.github.com/ptsiakos77/915816b097b4b088d5bd9d8022004e79

Thanks @ptsiakos77 for providing feedback with the message consumer @tuan-pham and I worked on 🙂 
I can verify in my local that your `from_term` suggestion is working 🙂  We are still working on putting up a message provider PR soon. Whichever you prefer, you may file this as a separate PR or we can include it as part of our message provider pr.

ptsiakos77 commented:
@williaminfante Thank you for your response! I am fine by including this as part of your message provider PR.
Creat job once again!

mefellows commented:
Wow, great to see this PR in use already!
